e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +223,77 @@ func (pf *PortForwarder) start() error {
 	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
 	go func() {
 		for {
-			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns, testwatcher.WatchTimeout(0), testwatcher.WatchContext(pf.ctx)); err != nil {
-				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
-				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
+			if err := pf.waitAndRestart(); err != nil {
+				pf.logger.Infof("PortForwarder finished: %v", err)
 				return
 			}
 		}
 	}()
 	return nil
 }
+
+// waitAndRestart will block until seeing an indicator that the port-forward is
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	// create an inner context for the watch so we can cancel threads between iterations
+	// the subprocess will be killed inside portForwardToDeployment
+	watchCtx, watchCancel := context.WithCancel(pf.ctx)
+	defer watchCancel()
+	// restart the port-forward if:
+	// - the subprocess is killed
+	// - the pod is restarted
+	restartChan := pf.signalOnFailure(
```

Comment:
The name "signalOnFailure" doesn't make sense here. Both Wait and Watch can return nil errors, which both need to trigger restart. So you're actually waiting for any one of these functions to return. And then once one of them returns you want to cancel the others. You'll need to pass a context to each function so they can be cancelled in a self-contained way without side effects. Then you can push the `WithCancel` setup into the `signalOnFailure` method. Then it makes sense to rename the method to something like `waitForAny` or even `race`, since it's a controlled race with only one winner.

Reply:
The "failure" I was intending to refer to here is the higher level failure of either:
- the subprocess exiting, i.e. "failing"
- the pod exiting, i.e. "failing"

The way we detect those failures is using blocking `Wait` statements, so I can understand the confusion. We can pass a context to `Watch` (which we do), however AFAIK we can't pass a context to `cmd.Wait()`. I think we could work around this by creating a cancellable context for the underlying `Command` and cancelling that here, which would ultimately lead to `cmd.Wait()` returning. I'll take a pass at refactoring